Added ina228 example

Louis31423142 opened the pull request.

Louis31423142 committed:
Added ina228 i2c example

Louis31423142 committed:
Added ina228 i2c example

lurch commented on i2c/ina228_i2c/README.md:
```diff
@@ -0,0 +1,37 @@
+[[bmp280_i2c_wiring]]
+[pdfwidth=75%]
+.Wiring Diagram for BMP280 sensor via I2C.
+image::bmp280_i2c_bb.png[]
```
Presumably this bit needs updating? :joy:

lurch commented on i2c/ina228_i2c/README.md:
```diff
@@ -0,0 +1,37 @@
+| *Item* | *Quantity* | Details
+| Breadboard | 1 | generic part
+| Raspberry Pi Pico 2 | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico-2/
+| INA228-based breakout board | 1 | https://www.adafruit.com/product/5832?srsltid=AfmBOoq8cjiUxn01ClOriCxHK3rMMtswYVhzd2tTkiLdjysj2CJZE-RX[from Adafruit]
```
Note that this is the AsciiDoc style link-format (see e.g. https://github.com/raspberrypi/pico-examples/tree/master/i2c/lcd_1602_i2c#readme ) whereas the file you're editing here has the MarkDown extension. Perhaps you need to rename this `README.md` to `README.adoc` ?

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+// LED GPIOs
+#define EXT_LED_GPIO 2
+
+// INA228 integrated shunt resistance
```
in Ohms?

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+// INA228 integrated shunt resistance
+#define R_SHUNT 0.015
+
+// Max expected current through shunt resistor
```
in Amps?

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+static void ina228_init() {
+    // Write to shunt register 
+    uint8_t shunt_msb = (SHUNT_CAL >> 8) & 0xFF;
+    uint8_t shunt_lsb = SHUNT_CAL & 0xff;
```
Minor nitpick, but it looks odd having `0xFF` on one line and then `0xff` on the next line :wink:

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+    uint8_t shunt_buf[3] = {SHUNT_CAL_REG, shunt_msb, shunt_lsb};
+    i2c_write_blocking(i2c_default, I2C_ADDR, shunt_buf, 3, false);
+
+    // Write to ADC config register (need to enable continuous mode to access accumilation variables energy and charge)
```
```suggestion
    // Write to ADC config register (need to enable continuous mode to access accumulation variables energy and charge)
```

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+
+static void ina228_read(float *vshunt, float *vbus, float *dietemp, float *current, float *power, float *energy, float *charge) {
+    // Buffers for writing register measurments to
+    // Some registers are 2 bytes, some are 3 bytes, the accumilation registers are 5 bytes
```
Same 'accumilation' typo as earlier.

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+    uint8_t buffer_3[3];
+    uint8_t buffer_5[5];
+
+    // For some of the registers the 4 least significant bits are reserved and always read zero, which is why we shift right 12, 4, then left 4 for vshunt, vbus and current
```
I'm not sure that you need so much detail here, it took me a while to understand what this comment was saying! Perhaps something like this is sufficient?
```suggestion
    // For the vshunt, vbus and current registers the 4 least significant bits are reserved and always read zero, so use bit-shifts to just extract the actual data
```

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+
+    i2c_write_blocking(i2c_default, I2C_ADDR, &POWER_REG, 1, true);
+    i2c_read_blocking(i2c_default, I2C_ADDR, buffer_3, 3, false);
+    *power = ((buffer_3[0] << 16) | (buffer_3[1] << 8) | buffer_3[2]) * POWER_FACTOR;
```
I wonder if creating some macros would make all this repeated bit-shifting more readable? E.g. you could add
```c
#define FLATTEN3(a3) ((a3[0] << 16) | (a3[1] << 8) | a3[2])
```
and then you could replace
```c
*power = ((buffer_3[0] << 16) | (buffer_3[1] << 8) | buffer_3[2]) * POWER_FACTOR;
```
with
```c
*power = FLATTEN3(buffer_3) * POWER_FACTOR;
```
? (with similar things being done for your `buffer_2` and `buffer_5` arrays).
Entirely up to you.

peterharperuk replied:
I'm not a big fan of the names as they don't really say what they're doing. Maybe inline functions would be better - and type safe (FLATTEN5 takes a uint64_t?). But I'm not sure it's worth more changes.

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+    while (true) {
+        ina228_read(&vshunt, &vbus, &dietemp, &current, &power, &energy, &charge);
+        printf("INA228 Measurments:\nVSHUNT: %f\nVBUS: %f\nDIETEMP: %f\nCURRENT: %f\nPOWER: %f\nENERGY: %.9f\nCHARGE: %.9f\n-----------------\n", vshunt, vbus, dietemp, current, power, energy, charge);
+        sleep_ms(3000);
```
3 seconds seems like a big sleep time, I think most of the other examples (that do things in a loop) sleep for one second?

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,119 @@
+
+    while (true) {
+        ina228_read(&vshunt, &vbus, &dietemp, &current, &power, &energy, &charge);
+        printf("INA228 Measurments:\nVSHUNT: %f\nVBUS: %f\nDIETEMP: %f\nCURRENT: %f\nPOWER: %f\nENERGY: %.9f\nCHARGE: %.9f\n-----------------\n", vshunt, vbus, dietemp, current, power, energy, charge);
```
IMHO would be useful to also display the units here.

lurch commented:
And if you could squash your commits to get rid of all the unrelated churn, that'd be great :+1: 

EDIT: Thanks :smiley:

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -14,7 +14,7 @@ The INA228 is highly configurable. Find the datasheet online (https://www.ti.com
 [[bmp280_i2c_wiring]]
```
Still says bmp280 here

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -14,7 +14,7 @@ The INA228 is highly configurable. Find the datasheet online (https://www.ti.com
 [[bmp280_i2c_wiring]]
 [pdfwidth=75%]
 .Wiring Diagram for BMP280 sensor via I2C.
```
...and BMP280 here

peterharperuk commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -0,0 +1,37 @@
+= Reading an INA228 power sensor via I2C
+
+This example code shows how to interface the Raspberry Pi Pico with the INA228 current/voltage/power/temperature/energy/charge sensor. The sensor features a 20 bit ADC and supports up to 85V DC, making it a great choice for high preccision measurments. It is capable of high-side and low-side measurments (which can be configured by soldering the jumper on the back of the board), but this example f ...
```
spelling preccision

peterharperuk replied:
We should probably include a link to the adafruit board https://thepihut.com/products/adafruit-ina228-i2c-85v-20-bit-high-or-low-side-power-monitor-stemma-qt-qwiic
And it should say somewhere that in this example we're measuring the power used by a led?

lurch replied:
Also 'measurments' -> 'measurements' (in several places!)
And IMHO "a 20 bit ADC" should be hyphenated as "a 20-bit ADC"

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -0,0 +1,37 @@
+| *Item* | *Quantity* | Details
+| Breadboard | 1 | generic part
+| Raspberry Pi Pico 2 | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico-2/
+| INA228-based breakout board | 1 | https://www.adafruit.com/product/5832?srsltid=AfmBOoq8cjiUxn01ClOriCxHK3rMMtswYVhzd2tTkiLdjysj2CJZE-RX
```
This URL can be shortened to just https://www.adafruit.com/product/5832

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,125 @@
+#include "hardware/i2c.h"
+#include <math.h>
+
+// LED GPIOs
```
```suggestion
// LED GPIO
```
:stuck_out_tongue_winking_eye:

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,125 @@
+// LED GPIOs
+#define EXT_LED_GPIO 2
+
+// INA228 integrated shunt resistance in Ohms
```
Is this shunt resistor actually part of the INA228, or just part of the Adafruit breakout board? If the latter, then IMHO the word "integrated" here ought to be removed.

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,125 @@
+#define FLATTEN2(a2) ((a2[0] << 8) + a2[1])
+#define FLATTEN3(a3) ((a3[0] << 16) | (a3[1] << 8) | a3[2])
+#define FLATTEN3_RESERVED(a3) ((a3[0] << 12) | (a3[1] << 4) | (a3[2] >> 4))
+#define FLATTEN5(a5) (((uint64_t) a5[0] << 32) | (a5[1] << 24) | (a5[2] << 16) | (a5[3] << 8) | a5[4])
```
@peterharperuk I'm not a "native C programmer" - does my suggestion of these macros make sense? Is there perhaps a better name for them? :man_shrugging:

peterharperuk replied:
My comment disappeared...

I'm not a big fan of the names as they don't really say what they're doing. Maybe inline functions would be better - and type safe (FLATTEN5 takes a uint64_t?). But I'm not sure it's worth more changes.

lurch replied:
@peterharperuk  Perhaps e.g. `FLATTEN2` could be renamed to `COALESCE_2_BYTES` ?

@Louis31423142 If you _wanted_ to use inline functions, I think you could do something like:
```c
static uint16_t coalesce_2_bytes(uint8_t* bytes) {
    return (bytes[0] << 8) + bytes[1];
}
```
(but I agree that it's probably not worth it)

peterharperuk replied:
fixed i think?

lurch commented on i2c/ina228_i2c/ina228_i2c.c:
```diff
@@ -0,0 +1,125 @@
+
+    while (true) {
+        ina228_read(&vshunt, &vbus, &dietemp, &current, &power, &energy, &charge);
+        printf("INA228 Measurments:\nVSHUNT: %f V\nVBUS: %f V\nDIETEMP: %f °C\nCURRENT: %f A\nPOWER: %f W\nENERGY: %f J\nCHARGE: %f C\n-----------------\n", vshunt, vbus, dietemp, current, power, energy, charge);
```
```suggestion
        printf("INA228 Measurements:\nVSHUNT: %f V\nVBUS: %f V\nDIETEMP: %f °C\nCURRENT: %f A\nPOWER: %f W\nENERGY: %f J\nCHARGE: %f C\n-----------------\n", vshunt, vbus, dietemp, current, power, energy, charge);
```
:laughing:

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -0,0 +1,37 @@
+| Breadboard | 1 | generic part
+| Raspberry Pi Pico 2 | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico-2/
+| INA228-based breakout board | 1 | https://www.adafruit.com/product/5832?srsltid=AfmBOoq8cjiUxn01ClOriCxHK3rMMtswYVhzd2tTkiLdjysj2CJZE-RX
+| M/M Jumper wires | 9 | generic part
```
Your updated diagram has 8 jumper wires?

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -0,0 +1,37 @@
+|===
+| *Item* | *Quantity* | Details
+| Breadboard | 1 | generic part
+| Raspberry Pi Pico 2 | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico-2/
```
Does it specifically need to be a Pico2, or would it work with either a Pico1 or a Pico2?

lurch commented on i2c/ina228_i2c/README.adoc:
```diff
@@ -0,0 +1,37 @@
+|===
+| *Item* | *Quantity* | Details
+| Breadboard | 1 | generic part
+| Raspberry Pi Pico (any) | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico-2/
```
Probably makes sense to update this URL too. https://www.raspberrypi.com/products/raspberry-pi-pico/

Although thinking about it, I wonder if something like this works, or if it's too wordy?
```suggestion
| Raspberry Pi Pico or Pico 2 (any model) | 1 | https://www.raspberrypi.com/products/raspberry-pi-pico/
```

WDYT @peterharperuk ?

peterharperuk committed:
Update example a bit

Reduce the output. Come up with a more realistic example monitoring
the memory usage of another device. Add USB logging.
Now matches ina260 example.

peterharperuk committed:
Update example a bit

Reduce the output. Come up with a more realistic example monitoring
the memory usage of another device. Add USB logging.
Now matches ina260 example.

peterharperuk committed:
Update readme

peterharperuk committed:
Update readme

peterharperuk committed:
Apply suggestion from @lurch

Co-authored-by: Andrew Scheller <lurch@durge.org>

peterharperuk committed:
Apply suggestion from @lurch

Co-authored-by: Andrew Scheller <lurch@durge.org>

peterharperuk reviewed (changes requested):
For some reason this isn't working right

peterharperuk committed:
Make ina228 work

peterharperuk committed:
Reduce average samples

Check for invalid results.

peterharperuk reviewed (dismissed):
This works well now

will-v-pi reviewed (dismissed):
I think the more the merrier - people will have bought different ina2xx boards over the years, so it's nice to have examples that work with lots of them.

Maybe a future job could be to combine them a bit, but for now separate examples is fine

peterharperuk committed:
Merge branch 'develop' into ina228
